packages/json-rpc-engine/src/JsonRpcEngine.ts

```diff
@@ -407,21 +428,36 @@ export class JsonRpcEngine extends SafeEventEmitter {
         jsonrpc: '2.0',
         error,
       });
-    }
-
-    // Handle notifications.
-    // We can't use isJsonRpcNotification here because that narrows callerReq to
-    // "never" after the if clause for unknown reasons.
-    if (this.#notificationHandler && !isJsonRpcRequest(callerReq)) {
+    } else if (
+      this.#notificationHandler &&
+      isJsonRpcNotification(callerReq) &&
+      !isJsonRpcRequest(callerReq)
+    ) {
       try {
         await this.#notificationHandler(callerReq);
-      } catch (error) {
+      } catch (error: unknown) {
         return callback(error);
       }
       return callback(null);
     }
+    return this.#handleRequest(callerReq, callback);
+  }
 
-    let error: JsonRpcEngineCallbackError = null;
+  /**
+   * Processes request object, and
+   * passes any error and response object to the given callback.
+   *
+   * Does not reject.
+   *
+   * @param callerReq - The request object from the caller.
+   * @param callback - The callback function.
+   * @returns Nothing.
+   */
+  async #handleRequest(
```

Comment:
Is there another name we could come up with than `handleRequest`? It's my understanding that `handle` also handles the request, so it might be difficult to see at a glance how this is different.

Or, if we don't need to use this method anywhere else right now, perhaps we don't extract it yet?

Reply:
Merging this back into `#handle` sounds like a good idea. Applied here: https://github.com/MetaMask/core/pull/3906/commits/8854416640c21867b5cbbbf35a22cbfaf5aeb21c.

The param type narrowing that I was trying to achieve with this refactor are accomplished further up in the file: https://github.com/MetaMask/core/pull/3906/files#diff-07a20aba14734ed38b0fbb7950cc14249fa7b9f836236734edef2001fb16359eR213-R243, mostly with type assertions.